CCEffectNode - Fix a bunch of sizing issues and add tests for them

thayerandrews opened the pull request:
- Refactor CCRenderTexture's create method into create and createTextureAndFboWithPixelSize. In CCEffectNode, override the create method so pixel size is computed correctly. This fixes effect node's parent relative sizing while leaving render texture the same (because render texture needs even more changes that I don't want to tackle right now).
- Override setContentSizeType on effect node so it also marks the content size as dirty when it changes.
- If content size has changed since the last visit, destroy any allocated resources and recreate them (as CCRenderTexture already does).
- Pass the content size in points to CCEffectRenderer from sprite and effect node.
- Add a bunch of effect node sizing tests.

**Update**:
- Fix handling of sprite color and opacity when combined with effects.
- Fix handling of clipping nodes as children of effect nodes.
- Fix effect node resizing when an effect node's size is parent relative and the size of a parent changes.
- Add tests for all of this.

Commit:
CCEffectNode - Fix parent relative sizing

It looks like this is broken for CCRenderTexture as well but I'm intentionally not fixing it there
with this change. I'll add that fix separately.

Commit:
CCEffectsTest - Add tests of effect node anchor point, sizing, and child positioning

Commit:
Merge branch 'develop' into thayer/effectNodeSizing

Commit:
CCEffectRenderer - Fix the content size passed to the effect renderer

It should be the content size in points not the raw content size. Thanks to Scott for pointing
this out.

Commit:
CCEffectNode - Fix effect node resizing

Because content size changes are handled in render texture's visit method and because effect node overrides visit,
changes to effect node content size were not being applied.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
CCEffectsTest - Update test label - No functional change

Commit:
CCEffectsTest - Add test of effects and sprite colors / opacity

Commit:
CCEffectsTest - Add another stacking test for sprite color / opacity + effects

Leave stitching enabled on the stack but disable it after the second effect. This means our stack [A, B, C]
becomes [ AB, C ] instead of [ ABC ] which is another good case to test.

Commit:
CCEffects - Fix mixing of sprite color / opacity with effects

Some effects handled colors and opacity properly, some did not. And none of them did the
right thing if multiple effects were stacked together without stitching. Color would either be
multiplied in multiple times or not at all.

This change introduces a new code snippet for effect function inputs that is used if the effect
function is the first one in a stack. Otherwise, the alternative snippet is used. The initial snippet
for all effects multiplies the fragment color by the texture and the alternate snippet just fetches
the texture.

In plain english this means that the input to an effect stack is the sprite's texture times its color
and we insure that the texture times color multiply does not happen again in the stack.

slembcke commented on cocos2d/CCRenderTexture.m:
```diff
@@ -169,12 +169,14 @@ -(id)init
 
 -(void)create
 {
+    CGSize pixelSize = CGSizeMake(_contentSize.width * _contentScale, _contentSize.height * _contentScale);
```
Should be self.contentSizeInPoints.width \* _contentScale ?

thayerandrews replied:
Yup. This should be fixed as well but I was leaving it out of this set of commits since I only wanted to change CCEffectNode here.

slembcke commented on cocos2d/CCRenderTexture.m:
```diff
@@ -234,6 +236,10 @@ -(void)create
 	CC_CHECK_GL_ERROR_DEBUG();
 	glPopGroupMarkerEXT();
 	
+    // XXX Thayer says: I think this is incorrect for any situations where the content
```
Thought about this a couple months ago and was sort of avoiding it until it was a problem. If the content size is parent relative then the content size in points can change at any time (even between frames) without any warning or event that can be caught. While somebody might set up a render texture this way, I'm not even sure if there is anything remotely correct we can do.

When put that way, I'm wondering if the only correct thing we can do is to make it an error to use anything but points for the content size type.

thayerandrews replied:
I see what you mean. If the content size type is parent relative, is it possible to determine the correct size during visit and compare this to the size that was last used to alloc the texture / FBO? If the two aren't equal then we destroy and recreate the texture and FBO prior to rendering the render texture / effect node's children? 

I'm guessing there's a situation (or two or three) that I'm overlooking where this won't work. And it would also mean comparing the sizes every frame to see if the texture reallocation needs to happen.

slembcke replied:
The problem is that the "auto-draw" functionality is pretty rarely used I think. Personally, I never noticed it until I was rewriting the class for 3.1. Before that I'd never heard it come up in the forums for instance. CCRenderTexture is primarily used by calling the begin/end methods explicitly from outside of a draw method, usually in update, and often only when content changes. Deferring the resize would mean that it have to invalidate the content the user just drew into it.

SpriteBuilder doesn't let you explicitly set the content size/type of a CCSprite. I'm not really sure what would happen if you did it in code. In both cases, I think it makes very little or no sense to allow parent relative content sizes at all.

thayerandrews replied:
Okay. That makes sense for CCRenderTexture. CCEffectNode actually only has auto-draw so I think the "check the current dimensions against the texture dimensions" fix would work there no?

For CCRenderTexture, could we add another mode so the options are:
- Manual draw
- Auto draw
- Manual draw but auto-draw if it is determined that the dimensions are invalid.

Or is it better to just leave CCRenderTexture along until version 4 when we have more leeway to change its behavior?

slembcke replied:
I'm leaning towards leaving it alone until v4. There are already a number of other problems that I don't really know what to do about without breaking the API or making it even more complicated. SpriteBuilder doesn't expose render textures (yet), and I don't know of any issues that have come up related to this yet.

thayerandrews replied:
Sounds good. I won't make any changes to CCRenderTexture and I'll talk to Viktor about whether or not he wants to keep these changes to CCEffectNode.

Commit:
CCEffectsTest - Add test of effect node parent resizing

Commit:
CCEffectNode - More parent relative sizing fixes

Keep track of the last allocated content size and compare this to the current content size.
If the two are different, destroy and reallocate the associated texture and FBO. This fixes
cases where the effect node's size is parent relative (and does not change) but the parent's
size does change.

Commit:
CCEffects - Explicitly set the strong and assign property attributes

Bill and Scott have pointed out that the absence of these attributes cause warnings in non-ARC code.

Commit:
CCEffectStack - Fix shader uniform management

This wasn't causing any problems yet but shader uniforms for effect stacks were not being
handled properly under all circumstances. After name mangling, the old code would construct
a uniform dictionary that was never used at render time. By render time, the effects would just
receive the default uniform dictionary from CCNode.

Commit:
CCEffectStack - Add some comments to the stitching code - No functional change

Commit:
CCEffectNode - Add initializers with the same parameters as CCRenderTexture

And add convenience constructors too.

Commit:
CCEffectsTest - Add a test of clipping combined with effects

A clipping node renders into an effect node with an attached pixellate effect.

Commit:
CCEffectNode - Fix CCClippingNodes as children of CCEffectNode

Even if it had a stencil buffer, CCEffectNode was only clearing the color buffer.
